File: Haskell/src/Jogo/Arquitetura.hs
module Jogo.Arquitetura where

-- Determinando um tipo "Coordenada"
type Coordenada = (Int, Int)

-- Estrutura do navio
data Navio = Navio {
    tipo :: String,
    tamanho :: Int,
    posicoes :: [Coordenada],
    partesAtingidas :: [Coordenada] 
} deriving (Show, Eq)

-- Definindo os tipos do Navio
naviosPadrao :: [Navio]
naviosPadrao =
  [ Navio "Porta-Avioes" 5 [] []
  , Navio "Encouracado" 4 [] []
  , Navio "Submarino" 3 [] []
  , Navio "Cruzador"   3 [] []
  , Navio "Destroyer"  2 [] []
  ]

-- Definindo a orientação do Navio
data Orientacao = H | V deriving (Show, Eq)

-- Definindo os "estados" de cada célula
data Celula = Agua 
            | ParteNavio 
            | Atingido 
            | Erro
            deriving (Eq)

-- Exibição da célula
exibicaoCelula :: Celula -> String
exibicaoCelula Agua = "🌊"
exibicaoCelula ParteNavio = "🚢"
exibicaoCelula Atingido = "✅"
exibicaoCelula Erro = "🔥"

-- Definindo o tamanho padrão do tabuleiro
tamanhoTabuleiro :: Int
tamanhoTabuleiro = 10

-- Determinando o tipo "Tabuleiro" como uma matriz de "Celula"
type Tabuleiro = [[Celula]]

-- Cria um Tabuleiro só com água (vazio)
criacaoTabuleiro :: Tabuleiro
criacaoTabuleiro = replicate tamanhoTabuleiro (replicate tamanhoTabuleiro Agua)

File: Prolog/src/jogo/arquitetura.pl
:- module(jogo_arquitetura, [
    celula/1,
    criacao_tabuleiro/1,
    exibicao_celula/2,
    navios_disponiveis/1,
    tamanho_tabuleiro/1
]).

celula(agua).
celula(parte_navio).
celula(atingido).
celula(erro).

exibicao_celula(agua, '🌊').
exibicao_celula(parte_navio, '🚢').
exibicao_celula(atingido, '🔥').
exibicao_celula(erro, '❌').

tamanho_tabuleiro(10).

% Estrutura: navio(Tipo, Tamanho, Posicoes, PartesAtingidas)
navios_disponiveis([
    navio('Porta-Avioes', 5, [], []),
    navio('Encouracado',  4, [], []),
    navio('Submarino',    3, [], []),
    navio('Cruzador',     3, [], []),
    navio('Destroyer',    2, [], [])
]).

criacao_tabuleiro(Tabuleiro) :-
    tamanho_tabuleiro(N),
    length(Linha, N),
    maplist(=(agua), Linha), % CORRIGIDO: Usa 'agua' para consistência
    length(Tabuleiro, N),
    maplist(=(Linha), Tabuleiro).